ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -540,6 +549,43 @@ private EndpointResponse handleScalablePushQuery(
     return EndpointResponse.ok(queryStreamWriter);
   }
 
+  private EndpointResponse handleStreamPullQuery(
+      final ImmutableAnalysis analysis,
+      final ServiceContext serviceContext,
+      final ConfiguredStatement<Query> configured,
+      final CompletableFuture<Void> connectionClosedFuture) {
+
+    final TransientQueryMetadata transientQueryMetadata = ksqlEngine
+        .createStreamPullQuery(
+            serviceContext,
+            configured,
+            false
+        );
+    localCommands.ifPresent(lc -> lc.write(transientQueryMetadata));
+
+    // Be sure to close the stream in the event of any error. It's safe to close it twice,
+    // which will happen if we call close inside the block, and then again when we exit the
+    // try-with-resources
+    try (QueryStreamWriter queryStreamWriter = new QueryStreamWriter(
+        transientQueryMetadata,
+        disconnectCheckInterval.toMillis(),
+        OBJECT_MAPPER,
+        connectionClosedFuture
+    )) {
+      ksqlEngine.waitForStreamPullQuery(
```

Comment:
One critical thing that occurs to me is that this will block until the query is complete, but before it's actually begun writing the output to the output stream.  I think this relies on buffering all of the rows in transientQueryMetadata.rowqueue.  If it were to fill up, I think it might deadlock.

Also, we want to be able to be able to stream back results and not hold it all in memory, otherwise we'd be fairly limited.  We probably need the logic that determines when the query is complete to be a separate thread in the background that eventually closes that queue when it's complete.  That's similar to how we handle this with pull queries at the moment.

Reply:
This is an excellent catch. Thank you.

After digging in again, I realized that the root of this evil was the way that I tried to wrap the waiting logic up into the engine and re-use the QueryStreamWriter from push queries. The "right" way to do it is to create a new xQueryStreamWriter (I called it `StreamPullQueryStreamWriter` for maximum confusion). Then, we can check for completion during writer's poll loop the way the pull queries do.